RFCs/FS-1090-Generic-struct-type-whose-fields-are-all-unmanaged-types-is-unmanaged.md

```diff
@@ -0,0 +1,288 @@
+# F# RFC FS-1088 - (Generic struct type whose fields are all unmanaged types is unmanaged)
+
+The design suggestion [Generic struct type whose fields are all unmanaged types is unmanaged](https://github.com/fsharp/fslang-suggestions/issues/692) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] Approved in principle
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/692)
+* [ ] Implementation: [In progress](https://github.com/dotnet/fsharp/pull/6064)
+
+# Summary
+
+[summary]: #summary
+
+Allow generic structs to be `unmanaged` at construction if all fields are `unmanaged` types.
+
+# Motivation
+
+[motivation]: #motivation
+
+* Make it easier to author low-level code.
+* Improve interoperability with C#.
+
+# Detailed design
+
+[design]: #detailed-design
+
+Currently, it is not possible to define a generic struct with the `unmanaged` constraint and substitute the type with another generic struct, even if its type is also constrained to be `unmanaged`.
+
+Consider the following code:
+
+```fsharp
+[<Struct>]
+type MyStruct(x: int, y: int) =
+    member _.X = x
+    member _.Y = y
+
+[<Struct>]
+type MyStructGeneric<'T when 'T: unmanaged>(x: 'T, y: 'T) =
+    member _.X = x
+    member _.Y = y
+
+[<Struct>]
+type MyStructGenericWithNoConstraint<'T>(x: 'T, y: 'T) =
+    member _.X = x
+    member _.Y = y
+
+[<Struct>]
+type Test<'T when 'T: unmanaged> =
+    val element: 'T
+
+let works = Test<int>()
+let works2 = Test<MyStruct>()
+
+// error FS0001: A generic construct requires that the type 'MyStructGeneric<int>' is an unmanaged type
+let error = Test<MyStructGeneric<int>>()
+
+// error FS0001: A generic construct requires that the type 'MyStructGenericWithNoConstraint<int>' is an unmanaged type
+let error = Test<MyStructGenericWithNoConstraint<int>>()
+```
+
+Prior to this RFC, the example above will fail to compile with:
+
+```less
+let error = Test<MyStructGeneric<int>>()
+  ------------^^^^^^^^^^^^^^^
+
+stdin(6,13): error FS0001: A generic construct requires that the type 'MyStructGeneric<int>' is an unmanaged type
+```
+
+This proposal aims to resolve this inconsistency by treating any generic struct type as `unmanaged` if all its fields are unmanaged.
+
+__NOTE:__ Both generic struct types _with_ and _without_ the `unmanaged` constraint on type parameter(s) can be treated as unmanaged as long as type parameters are being substituted by an unmanaged type(s).
+
+For example, consider the following code:
+
+```fsharp
+[<Struct>]
+type MyStructGeneric<'T when 'T: unmanaged>(x: 'T, y: 'T) =
+    member _.X = x
+    member _.Y = y
+
+[<Struct>]
+type MyStructGenericWithNoConstraint<'T>(x: 'T, y: 'T) =
+    member _.X = x
+    member _.Y = y
+```
+
+Instances of both `MyStructGeneric<'T>` and `MyStructGenericWithNoConstraint<'T>` will be treated as unmanaged type as long as `'T` is being unmanaged.
```

Comment:
I think I understand this section, but it's a little confusing (perhaps due to my feedback).

Both of these struct types should be usable as an unmanaged type. That is, I could do this:

```fsharp
let t = Test<MyStructGeneric<int>>()
let t' = Test<MyStructGenericWithNoConstraint<int>>()
```

Having the constraint on the type I am substituting the `'T` in `Test` with is not required with this change.

In other words, any generic struct where all its fields are known to be unmanaged can be considered unmanaged.

Kind of a mouthful and getting the definition across clearly is kind of difficult. But I think if you include examples of instantiating `Test` with both it will be clearer.

Reply:
I've changed some wording, hope it's more clear now :)